NDTensors/ext/NDTensorsCUDAExt/set_types.jl

```diff
@@ -1,57 +1,21 @@
-# `SetParameters.jl` overloads.
-get_parameter(::Type{<:CuArray{P1}}, ::Position{1}) where {P1} = P1
-get_parameter(::Type{<:CuArray{<:Any,P2}}, ::Position{2}) where {P2} = P2
-get_parameter(::Type{<:CuArray{<:Any,<:Any,P3}}, ::Position{3}) where {P3} = P3
-
-# Set parameter 1
-set_parameter(::Type{<:CuArray}, ::Position{1}, P1) = CuArray{P1}
-set_parameter(::Type{<:CuArray{<:Any,P2}}, ::Position{1}, P1) where {P2} = CuArray{P1,P2}
-function set_parameter(::Type{<:CuArray{<:Any,<:Any,P3}}, ::Position{1}, P1) where {P3}
-  return CuArray{P1,<:Any,P3}
-end
-function set_parameter(::Type{<:CuArray{<:Any,P2,P3}}, ::Position{1}, P1) where {P2,P3}
-  return CuArray{P1,P2,P3}
-end
-
-# Set parameter 2
-set_parameter(::Type{<:CuArray}, ::Position{2}, P2) = CuArray{<:Any,P2}
-set_parameter(::Type{<:CuArray{P1}}, ::Position{2}, P2) where {P1} = CuArray{P1,P2}
-function set_parameter(::Type{<:CuArray{<:Any,<:Any,P3}}, ::Position{2}, P2) where {P3}
-  return CuArray{<:Any,P2,P3}
-end
-function set_parameter(::Type{<:CuArray{P1,<:Any,P3}}, ::Position{2}, P2) where {P1,P3}
-  return CuArray{P1,P2,P3}
-end
-
-# Set parameter 3
-set_parameter(::Type{<:CuArray}, ::Position{3}, P3) = CuArray{<:Any,<:Any,P3}
-set_parameter(::Type{<:CuArray{P1}}, ::Position{3}, P3) where {P1} = CuArray{P1,<:Any,P3}
-function set_parameter(::Type{<:CuArray{<:Any,P2}}, ::Position{3}, P3) where {P2}
-  return CuArray{<:Any,P2,P3}
-end
-set_parameter(::Type{<:CuArray{P1,P2}}, ::Position{3}, P3) where {P1,P2} = CuArray{P1,P2,P3}
-
-default_parameter(::Type{<:CuArray}, ::Position{1}) = Float64
-default_parameter(::Type{<:CuArray}, ::Position{2}) = 1
-default_parameter(::Type{<:CuArray}, ::Position{3}) = Mem.DeviceBuffer
-
-nparameters(::Type{<:CuArray}) = Val(3)
-
-SetParameters.unspecify_parameters(::Type{<:CuArray}) = CuArray
-
-using NDTensors.TypeParameterAccessors: TypeParameterAccessors
+# TypeParameterAccessors definitions
+using NDTensors.TypeParameterAccessors: TypeParameterAccessors, Position, set_type_parameter
 using NDTensors.GPUArraysCoreExtensions: storagemode
 ## TODO remove TypeParameterAccessors when SetParameters is removed
 function TypeParameterAccessors.position(::Type{<:CuArray}, ::typeof(eltype))
-  return TypeParameterAccessors.Position(1)
+  return Position(1)
 end
 function TypeParameterAccessors.position(::Type{<:CuArray}, ::typeof(Base.ndims))
-  return TypeParameterAccessors.Position(2)
+  return Position(2)
 end
 function TypeParameterAccessors.position(::Type{<:CuArray}, ::typeof(storagemode))
-  return TypeParameterAccessors.Position(3)
+  return Position(3)
 end
 
 function TypeParameterAccessors.default_type_parameters(::Type{<:CuArray})
   return (Float64, 1, CUDA.Mem.DeviceBuffer)
 end
+
```

Comment:
Because there is no `set_ndims(::Type{<:AbstractArray}` I defined one for `CuArray` and `MtlArray`. Potentially that is not right? I also have not yet moved `NDTensors.ndims` to `TypeParameterAccessors`

Reply:
I think this can be turned into a generic `AbstractArray` function based on the way it is written here.

Reply:
I guess one thing to think about with this is what happens when you want to do 
```
set_ndims(Transpose{Float32, Matrix{Float32}}, 3)
```
Should ndims be based on `IsWrappedArray`? But then an issue I have had before is that `TensorStorage` is a wrapped array so `set_ndims` of a tensor storage will try to set the dimension of the inner data which right now is required to be a vector.

Reply:
`set_ndims` for `Transpose` will fail since it doesn't have an `ndims` type parameter. I don't think we have to worry about `set_ndims` for wrapped array types right now.

Reply:
Oh good point, I forgot `Transpose` doesn't have `position(::, typeof(ndims))` defined. Thanks!

Reply:
It's a good question to think about though, my guess is that `set_ndims` just won't really make sense for most wrapper array types, so we might consider constraining the generic definition to `!IsWrappedArray`.